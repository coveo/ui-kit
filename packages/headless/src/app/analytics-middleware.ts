import {Middleware} from 'redux';
import {getRecommendations} from '../features/recommendation/recommendation-actions';
import {executeSearch} from '../features/search/search-actions';
import {getProductRecommendations} from '../features/product-recommendations/product-recommendations-actions';

export const analyticsMiddleware: Middleware = (api) => (next) => (action) => {
  // Why all these shenanigans ?
  // https://redux.js.org/style-guide/style-guide#do-not-put-non-serializable-values-in-state-or-actions
  // analyticsAction is returned as a function in some action payload. A function is non serializable.
  // This is not recommended, with one exception: It has to be handled before any reducer has the chance to do it's job, and should not stored in the state.
  // To meet those conditions, we ensure the analyticsMiddleware is always the first executed (in store.ts/configureStore) and that it is removed from the payload.
  // We only keep a reference to the function in order to dispatch it correctly after the reducer (next) has the chance of processing the action/state change
  const analytics = action.payload?.analyticsAction;
  if (analytics !== undefined) {
    delete action.payload?.analyticsAction;
  }

  const ret = next(action);

  if (action.type === executeSearch.fulfilled.type && analytics === undefined) {
    console.error('No analytics action associated with search:', action);
  }

  if (action.type === getRecommendations.fulfilled && analytics === undefined) {
    console.error(
      'No analytics action associated with recommendation:',
      action
    );
  }

  if (
    action.type === getProductRecommendations.fulfilled &&
    analytics === undefined
  ) {
    console.error(
      'No analytics action associated with product recommendation:',
      action
    );
  }

  if (analytics !== undefined) {
    api.dispatch(analytics);
  }

  return ret;
};
