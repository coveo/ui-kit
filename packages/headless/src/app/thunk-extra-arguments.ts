import {AnalyticsClientSendEventHook} from 'coveo.analytics';
import {Logger} from 'pino';
import {PreprocessRequest} from '../api/preprocess-request';
import {validatePayloadAndThrow} from '../utils/validate-payload';
import {EngineConfiguration} from './engine-configuration';
import {NoopPreprocessRequest} from '../api/preprocess-request';

export interface ThunkExtraArguments {
  preprocessRequest?: PreprocessRequest;
  analyticsClientMiddleware: AnalyticsClientSendEventHook;
  logger: Logger;
  validatePayload: typeof validatePayloadAndThrow;
}

export function buildThunkExtraArguments(
  configuration: EngineConfiguration,
  logger: Logger
): ThunkExtraArguments {
  const analyticsClientMiddleware = getAnalyticsClientMiddleware(configuration);
  const validatePayload = validatePayloadAndThrow;
  const preprocessRequest = getPreprocessRequest(configuration);

  return {
    analyticsClientMiddleware,
    validatePayload,
    preprocessRequest,
    logger,
  };
}

function getAnalyticsClientMiddleware(
  configuration: EngineConfiguration
): AnalyticsClientSendEventHook {
  const {analytics} = configuration;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const NoopAnalyticsMiddleware = (_: string, p: any) => p;
  return analytics?.analyticsClientMiddleware || NoopAnalyticsMiddleware;
}

function getPreprocessRequest(configuration: EngineConfiguration) {
  return configuration.preprocessRequest || NoopPreprocessRequest;
}
