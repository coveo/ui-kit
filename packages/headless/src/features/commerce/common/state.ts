type SliceId = string;
export type Sliced<Slice> = Record<SliceId, Slice>;

export type CommerceSlicesState = Sliced<unknown>;

export function getSliced<
  Slice,
>(): Sliced<Slice> {
  return {};
}
