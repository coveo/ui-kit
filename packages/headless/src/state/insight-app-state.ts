import {
  ConfigurationSection,
  InsightConfigurationSection,
  InsightInterfaceSection,
  SearchHubSection,
  VersionSection,
  InsightCaseContextSection,
  FacetSection,
  QuerySection,
  PaginationSection,
  ResultPreviewSection,
  SearchSection,
  FacetSearchSection,
  FacetOptionsSection,
  QuerySuggestionSection,
  QuerySetSection,
  NumericFacetSection,
  DateFacetSection,
  CategoryFacetSection,
  SortSection,
  StaticFilterSection,
  DidYouMeanSection,
  TabSection,
  FieldsSection,
  AttachedResultsSection,
  QuestionAnsweringSection,
  FoldingSection,
  GeneratedAnswerSection,
  ContextSection,
} from './state-sections';

export type InsightSearchParametersState = FacetSection &
  DateFacetSection &
  NumericFacetSection &
  CategoryFacetSection &
  QuerySection &
  TabSection &
  FieldsSection &
  SortSection;

export type InsightAppState = InsightSearchParametersState &
  SearchHubSection &
  ConfigurationSection &
  VersionSection &
  InsightConfigurationSection &
  SearchSection &
  PaginationSection &
  InsightInterfaceSection &
  InsightCaseContextSection &
  ResultPreviewSection &
  FacetSection &
  FacetSearchSection &
  FacetOptionsSection &
  QuerySetSection &
  QuerySuggestionSection &
  SortSection &
  StaticFilterSection &
  DidYouMeanSection &
  AttachedResultsSection &
  QuestionAnsweringSection &
  FoldingSection &
  GeneratedAnswerSection &
  ContextSection;
