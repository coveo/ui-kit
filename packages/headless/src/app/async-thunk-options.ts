import {ThunkExtraArguments} from './thunk-extra-arguments';

export interface AsyncThunkOptions<
  T,
  TExtra extends ThunkExtraArguments = ThunkExtraArguments,
> {
  state: T;
  extra: TExtra;
}
