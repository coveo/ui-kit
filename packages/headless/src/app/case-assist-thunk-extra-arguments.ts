import {CaseAssistAPIClient} from '../api/service/case-assist/case-assist-api-client.js';
import {ClientThunkExtraArguments} from './thunk-extra-arguments.js';

export interface CaseAssistThunkExtraArguments
  extends ClientThunkExtraArguments<CaseAssistAPIClient> {}
