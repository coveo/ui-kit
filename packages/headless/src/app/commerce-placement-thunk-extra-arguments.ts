import {CommerceUnifiedAPIClient} from '../api/commerce/unified-api/unified-api-client';
import {ClientThunkExtraArguments} from './thunk-extra-arguments';

export interface CommercePlacementsThunkExtraArguments
  extends ClientThunkExtraArguments<CommerceUnifiedAPIClient> {}
