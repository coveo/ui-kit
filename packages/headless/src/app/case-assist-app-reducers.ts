import {ReducersMapObject} from 'redux';
import {CaseAssistAppState} from '../state/case-assist-app-state';
import {caseAssistReducer} from '../features/case-assist/case-assist-slice';
import {configurationReducer} from '../features/configuration/configuration-slice';

export const caseAssistAppReducers: ReducersMapObject<CaseAssistAppState> = {
  configuration: configurationReducer,
  caseAssist: caseAssistReducer,
};
