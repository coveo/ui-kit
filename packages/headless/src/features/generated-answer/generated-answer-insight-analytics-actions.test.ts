import {ThunkExtraArguments} from '../../app/thunk-extra-arguments';
import {InsightAppState} from '../../state/insight-app-state';
import {
  buildMockInsightEngine,
  MockedInsightEngine,
} from '../../test/mock-engine-v2';
import {buildMockInsightState} from '../../test/mock-insight-state';
import {
  logGeneratedAnswerDetailedFeedback,
  logGeneratedAnswerFeedback,
  logRetryGeneratedAnswer,
  logRephraseGeneratedAnswer,
  logOpenGeneratedAnswerSource,
  logHoverCitation,
  logLikeGeneratedAnswer,
  logDislikeGeneratedAnswer,
  logGeneratedAnswerStreamEnd,
  logGeneratedAnswerShowAnswers,
  logGeneratedAnswerHideAnswers,
  logCopyGeneratedAnswer,
} from './generated-answer-insight-analytics-actions';

const mockLogGeneratedAnswerFeedbackSubmit = jest.fn();
const mockLogRetryGeneratedAnswer = jest.fn();
const mockLogRephraseGeneratedAnswer = jest.fn();
const mockLogOpenGeneratedAnswerSource = jest.fn();
const mockLogHoverCitation = jest.fn();
const mockLogLikeGeneratedAnswer = jest.fn();
const mockLogDislikeGeneratedAnswer = jest.fn();
const mockLogGeneratedAnswerStreamEnd = jest.fn();
const mockLogGeneratedAnswerShowAnswers = jest.fn();
const mockLogGeneratedAnswerHideAnswers = jest.fn();
const mockLogCopyGeneratedAnswer = jest.fn();

jest.mock('coveo.analytics', () => {
  const mockCoveoInsightClient = jest.fn(() => ({
    disable: jest.fn(),
    logGeneratedAnswerFeedbackSubmit: mockLogGeneratedAnswerFeedbackSubmit,
    logRetryGeneratedAnswer: mockLogRetryGeneratedAnswer,
    logRephraseGeneratedAnswer: mockLogRephraseGeneratedAnswer,
    logOpenGeneratedAnswerSource: mockLogOpenGeneratedAnswerSource,
    logGeneratedAnswerSourceHover: mockLogHoverCitation,
    logLikeGeneratedAnswer: mockLogLikeGeneratedAnswer,
    logDislikeGeneratedAnswer: mockLogDislikeGeneratedAnswer,
    logGeneratedAnswerStreamEnd: mockLogGeneratedAnswerStreamEnd,
    logGeneratedAnswerShowAnswers: mockLogGeneratedAnswerShowAnswers,
    logGeneratedAnswerHideAnswers: mockLogGeneratedAnswerHideAnswers,
    logGeneratedAnswerCopyToClipboard: mockLogCopyGeneratedAnswer,
  }));

  return {
    CoveoInsightClient: mockCoveoInsightClient,
    history: {HistoryStore: jest.fn()},
  };
});

const exampleFeedback = 'irrelevant';
const exampleGenerativeQuestionAnsweringId = '123';
const exampleDetails = 'example details';
const exampleCitationId = 'citation_id';
const exampleCitationPermanentid = 'citation_permanentid';
const exampleSubject = 'example subject';
const exampleDescription = 'example description';
const exampleCaseId = '1234';
const exampleCaseNumber = '5678';

const expectedCaseContext = {
  caseContext: {
    Case_Subject: exampleSubject,
    Case_Description: exampleDescription,
  },
  caseId: exampleCaseId,
  caseNumber: exampleCaseNumber,
};

describe('the analytics related to the generated answer feature in the insight use case', () => {
  let engine: MockedInsightEngine;
  let state: InsightAppState;

  beforeEach(() => {
    state = buildMockInsightState();
    state.search.response.extendedResults.generativeQuestionAnsweringId =
      exampleGenerativeQuestionAnsweringId;
    state.generatedAnswer.citations = [
      {
        id: exampleCitationId,
        permanentid: exampleCitationPermanentid,
        title: 'example title',
        uri: 'example: uri',
      },
    ];
    state.insightCaseContext = {
      caseContext: {
        Case_Subject: exampleSubject,
        Case_Description: exampleDescription,
      },
      caseId: exampleCaseId,
      caseNumber: exampleCaseNumber,
    };
    engine = buildMockInsightEngine(state);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should log #logRetryGeneratedAnswer with the right payload', async () => {
    await logRetryGeneratedAnswer()()(
      engine.dispatch,
      () => engine.state,
      {} as ThunkExtraArguments
    );

    const mockToUse = mockLogRetryGeneratedAnswer;

    expect(mockToUse).toHaveBeenCalledTimes(1);
    expect(mockToUse).toHaveBeenCalledWith(expectedCaseContext);
  });

  it('should log #logRephraseGeneratedAnswer with the right payload', async () => {
    const exampleRephraseFormat = 'step';
    await logRephraseGeneratedAnswer({answerStyle: exampleRephraseFormat})()(
      engine.dispatch,
      () => engine.state,
      {} as ThunkExtraArguments
    );

    const mockToUse = mockLogRephraseGeneratedAnswer;
    const expectedMetadata = {
      generativeQuestionAnsweringId: exampleGenerativeQuestionAnsweringId,
      rephraseFormat: exampleRephraseFormat,
    };

    expect(mockToUse).toHaveBeenCalledTimes(1);
    expect(mockToUse).toHaveBeenCalledWith(
      expectedMetadata,
      expectedCaseContext
    );
  });

  it('should log #logOpenGeneratedAnswerSource with the right payload', async () => {
    await logOpenGeneratedAnswerSource(exampleCitationId)()(
      engine.dispatch,
      () => engine.state,
      {} as ThunkExtraArguments
    );

    const mockToUse = mockLogOpenGeneratedAnswerSource;
    const expectedMetadata = {
      generativeQuestionAnsweringId: exampleGenerativeQuestionAnsweringId,
      permanentId: exampleCitationPermanentid,
      citationId: exampleCitationId,
    };

    expect(mockToUse).toHaveBeenCalledTimes(1);
    expect(mockToUse).toHaveBeenCalledWith(
      expectedMetadata,
      expectedCaseContext
    );
  });

  it('should log #logHoverCitation with the right payload', async () => {
    const exampleHoverTime = 100;
    await logHoverCitation(exampleCitationId, exampleHoverTime)()(
      engine.dispatch,
      () => engine.state,
      {} as ThunkExtraArguments
    );

    const mockToUse = mockLogHoverCitation;
    const expectedMetadata = {
      generativeQuestionAnsweringId: exampleGenerativeQuestionAnsweringId,
      permanentId: exampleCitationPermanentid,
      citationId: exampleCitationId,
      citationHoverTimeMs: exampleHoverTime,
    };

    expect(mockToUse).toHaveBeenCalledTimes(1);
    expect(mockToUse).toHaveBeenCalledWith(
      expectedMetadata,
      expectedCaseContext
    );
  });

  it('should log #logLikeGeneratedAnswer with the right payload', async () => {
    await logLikeGeneratedAnswer()()(
      engine.dispatch,
      () => engine.state,
      {} as ThunkExtraArguments
    );

    const mockToUse = mockLogLikeGeneratedAnswer;
    const expectedMetadata = {
      generativeQuestionAnsweringId: exampleGenerativeQuestionAnsweringId,
    };

    expect(mockToUse).toHaveBeenCalledTimes(1);
    expect(mockToUse).toHaveBeenCalledWith(
      expectedMetadata,
      expectedCaseContext
    );
  });

  it('should log #logDislikeGeneratedAnswer with the right payload', async () => {
    await logDislikeGeneratedAnswer()()(
      engine.dispatch,
      () => engine.state,
      {} as ThunkExtraArguments
    );

    const mockToUse = mockLogDislikeGeneratedAnswer;
    const expectedMetadata = {
      generativeQuestionAnsweringId: exampleGenerativeQuestionAnsweringId,
    };

    expect(mockToUse).toHaveBeenCalledTimes(1);
    expect(mockToUse).toHaveBeenCalledWith(
      expectedMetadata,
      expectedCaseContext
    );
  });

  it('should log #logGeneratedAnswerFeedback with the right payload', async () => {
    await logGeneratedAnswerFeedback(exampleFeedback)()(
      engine.dispatch,
      () => engine.state,
      {} as ThunkExtraArguments
    );

    const mockToUse = mockLogGeneratedAnswerFeedbackSubmit;
    const expectedMetadata = {
      generativeQuestionAnsweringId: exampleGenerativeQuestionAnsweringId,
      reason: exampleFeedback,
    };

    expect(mockToUse).toHaveBeenCalledTimes(1);
    expect(mockToUse).toHaveBeenCalledWith(
      expectedMetadata,
      expectedCaseContext
    );
  });

  it('should log #logGeneratedAnswerDetailedFeedback with the right payload', async () => {
    await logGeneratedAnswerDetailedFeedback(exampleDetails)()(
      engine.dispatch,
      () => engine.state,
      {} as ThunkExtraArguments
    );

    const mockToUse = mockLogGeneratedAnswerFeedbackSubmit;
    const expectedMetadata = {
      generativeQuestionAnsweringId: exampleGenerativeQuestionAnsweringId,
      reason: 'other',
      details: exampleDetails,
    };

    expect(mockToUse).toHaveBeenCalledTimes(1);
    expect(mockToUse).toHaveBeenCalledWith(
      expectedMetadata,
      expectedCaseContext
    );
  });

  it('should log #logGeneratedAnswerStreamEnd with the right payload', async () => {
    await logGeneratedAnswerStreamEnd(true)()(
      engine.dispatch,
      () => engine.state,
      {} as ThunkExtraArguments
    );

    const mockToUse = mockLogGeneratedAnswerStreamEnd;
    const expectedMetadata = {
      generativeQuestionAnsweringId: exampleGenerativeQuestionAnsweringId,
      answerGenerated: true,
    };

    expect(mockToUse).toHaveBeenCalledTimes(1);
    expect(mockToUse).toHaveBeenCalledWith(
      expectedMetadata,
      expectedCaseContext
    );
  });

  it('should log #logGeneratedAnswerShowAnswers with the right payload', async () => {
    await logGeneratedAnswerShowAnswers()()(
      engine.dispatch,
      () => engine.state,
      {} as ThunkExtraArguments
    );

    const mockToUse = mockLogGeneratedAnswerShowAnswers;
    const expectedMetadata = {
      generativeQuestionAnsweringId: exampleGenerativeQuestionAnsweringId,
    };

    expect(mockToUse).toHaveBeenCalledTimes(1);
    expect(mockToUse).toHaveBeenCalledWith(
      expectedMetadata,
      expectedCaseContext
    );
  });

  it('should log #logGeneratedAnswerHideAnswers with the right payload', async () => {
    await logGeneratedAnswerHideAnswers()()(
      engine.dispatch,
      () => engine.state,
      {} as ThunkExtraArguments
    );

    const mockToUse = mockLogGeneratedAnswerHideAnswers;
    const expectedMetadata = {
      generativeQuestionAnsweringId: exampleGenerativeQuestionAnsweringId,
    };

    expect(mockToUse).toHaveBeenCalledTimes(1);
    expect(mockToUse).toHaveBeenCalledWith(
      expectedMetadata,
      expectedCaseContext
    );
  });

  it('should log #logCopyGeneratedAnswer with the right payload', async () => {
    await logCopyGeneratedAnswer()()(
      engine.dispatch,
      () => engine.state,
      {} as ThunkExtraArguments
    );

    const mockToUse = mockLogCopyGeneratedAnswer;
    const expectedMetadata = {
      generativeQuestionAnsweringId: exampleGenerativeQuestionAnsweringId,
    };

    expect(mockToUse).toHaveBeenCalledTimes(1);
    expect(mockToUse).toHaveBeenCalledWith(
      expectedMetadata,
      expectedCaseContext
    );
  });
});
