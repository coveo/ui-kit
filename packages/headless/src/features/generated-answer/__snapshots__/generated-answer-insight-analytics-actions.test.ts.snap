// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generated answer insight analytics actions when analyticsMode is \`next\` should log #logCopyGeneratedAnswer with the right payload 1`] = `
[
  "Qna.AnswerAction",
  {
    "action": "copyToClipboard",
    "answer": {
      "responseId": "456",
      "type": "RGA",
    },
  },
]
`;

exports[`generated answer insight analytics actions when analyticsMode is \`next\` should log #logDislikeGeneratedAnswer with the right payload 1`] = `
[
  "Qna.SubmitFeedback",
  {
    "answer": {
      "responseId": "456",
      "type": "RGA",
    },
    "feedback": {
      "liked": false,
    },
  },
]
`;

exports[`generated answer insight analytics actions when analyticsMode is \`next\` should log #logGeneratedAnswerCollapse with the right payload 1`] = `
[
  "Qna.AnswerAction",
  {
    "action": "collapse",
    "answer": {
      "responseId": "456",
      "type": "RGA",
    },
  },
]
`;

exports[`generated answer insight analytics actions when analyticsMode is \`next\` should log #logGeneratedAnswerDetailedFeedback with the right payload 1`] = `
[
  "Qna.SubmitFeedback",
  {
    "answer": {
      "responseId": "456",
      "type": "RGA",
    },
    "feedback": {
      "details": "example details",
      "liked": false,
      "reason": "other",
    },
  },
]
`;

exports[`generated answer insight analytics actions when analyticsMode is \`next\` should log #logGeneratedAnswerExpand with the right payload 1`] = `
[
  "Qna.AnswerAction",
  {
    "action": "expand",
    "answer": {
      "responseId": "456",
      "type": "RGA",
    },
  },
]
`;

exports[`generated answer insight analytics actions when analyticsMode is \`next\` should log #logGeneratedAnswerHideAnswers with the right payload 1`] = `
[
  "Qna.AnswerAction",
  {
    "action": "hide",
    "answer": {
      "responseId": "456",
      "type": "RGA",
    },
  },
]
`;

exports[`generated answer insight analytics actions when analyticsMode is \`next\` should log #logGeneratedAnswerShowAnswers with the right payload 1`] = `
[
  "Qna.AnswerAction",
  {
    "action": "show",
    "answer": {
      "responseId": "456",
      "type": "RGA",
    },
  },
]
`;

exports[`generated answer insight analytics actions when analyticsMode is \`next\` should log #logHoverCitation with the right payload 1`] = `
[
  "Qna.CitationHover",
  {
    "answer": {
      "responseId": "456",
      "type": "RGA",
    },
    "citation": {
      "id": "citation_id",
      "type": "Source",
    },
    "citationHoverTimeInMs": 100,
  },
]
`;

exports[`generated answer insight analytics actions when analyticsMode is \`next\` should log #logLikeGeneratedAnswer with the right payload 1`] = `
[
  "Qna.SubmitFeedback",
  {
    "answer": {
      "responseId": "456",
      "type": "RGA",
    },
    "feedback": {
      "liked": true,
    },
  },
]
`;

exports[`generated answer insight analytics actions when analyticsMode is \`next\` should log #logOpenGeneratedAnswerSource with the right payload 1`] = `
[
  "Qna.CitationClick",
  {
    "answer": {
      "responseId": "456",
      "type": "RGA",
    },
    "citation": {
      "id": "citation_id",
      "type": "Source",
    },
  },
]
`;
